source/arm9/hw_math.c

```diff
@@ -0,0 +1,35 @@
+
+
+#include <math.h>
+
+
+
+f32 hw_sqrtf(f32 x){
+    union{f32 f; u32 i;}xu;
+    xu.f=x;
+    u64 mantissa=xu.i & ((1<<23)-1);
+    //check if exponent is odd
+    //before subtracting 127 exponent was even if it is odd now
+    //therefore check if last digit is 0 
+    mantissa=(mantissa+(1<<23))<<   ( (( xu.i & (1<<23))==0  )      +25 );
+    REG_SQRTCNT = SQRT_64;
+    REG_SQRT_PARAM = mantissa;
+    u32 raw_exponent= (xu.i & (0xff<<23));
+    if (raw_exponent > 0 ) {
+	s32 exponent=raw_exponent-(127<<23);
+	exponent=exponent>>1; //right shift on negative number depends on compiler
+	exponent=((exponent+(127<<23))& (0xff<<23) );
+    //fetch async result here
+	while(REG_SQRTCNT & SQRT_BUSY);
+	u32 new_mantissa=REG_SQRT_RESULT;
+	new_mantissa+=1;
+	new_mantissa>>=1;
+	xu.i=  exponent| (new_mantissa & ((1<<23)-1));
+	return xu.f;
+    } else{
+	while(REG_SQRTCNT & SQRT_BUSY);
```

Comment:
Do you really need to wait for the hardware to not be busy? It's probably okay to leave it running. If the user wants to start a new operation they will overwrite the values in the registers, they don't have to wait until the hardware is done (I may be wrong here, though).

Reply:
I was under the impression that REG_SQRCNT can be both written and read to, and so if the sqrt unit changes its value after the first sqrt is done it might lead to fetching the result of the previous sqrt operation.

Reply:
Hmmm well, I don't know, let's leave it like this for now, it won't hurt.